cql3/statements/ks_prop_defs.cc

```diff
@@ -68,6 +68,39 @@ static std::map<sstring, sstring> prepare_options(
         }
     }
 
+    bool auto_initial_tablets = false;
+    if (feat.tablets) {
+        auto it = options.find(ks_prop_defs::INITIAL_TABLETS_KEY);
+        if (it == options.end()) {
+            auto_initial_tablets = true;
+        }
+    }
+
+    if (auto_initial_tablets) {
+        const auto& topo = tm.get_topology();
+        unsigned initial_tablets = std::numeric_limits<unsigned>::min();
+        for (const auto& dc : topo.get_datacenter_endpoints()) {
+            unsigned shards_in_dc = 0;
+            unsigned rf_in_dc = 1;
+
+            for (const auto& ep : dc.second) {
+                const auto* node = topo.find_node(ep);
+                if (node != nullptr) {
+                    shards_in_dc += node->get_shard_count();
+                }
+            }
+
+            if (auto it = options.find(dc.first); it != options.end()) {
+                rf_in_dc = std::stol(it->second);
+            }
+
+            unsigned tablets_in_dc = (shards_in_dc + rf_in_dc - 1) / rf_in_dc;
+            initial_tablets = std::max(initial_tablets, tablets_in_dc);
+        }
+
+        options.emplace(ks_prop_defs::INITIAL_TABLETS_KEY, fmt::to_string(initial_tablets));
+    }
```

Comment:
It would be better to decide this on tablet allocation time, when table is created (in network_topology_strategy::allocate_tablets_for_new_table). The topology may have changed since the keyspace was created.

at keyspace creation time, tablet_aware_replication_strategy::process_tablet_options() should just set _uses_tablets = true when the feature is enabled.

I think we shouldn't set "initial_tablets" on behalf of the user.

Reply:
Not that I disagree, but the same applies to per-DC replication factors and they _are_ set here if not supplied by user. Why are tablets different in this sense?

Reply:
With `initial_tablets` it's more harmful, since the user's intent is not to use a specific tablet count, while with RF he specified the intent to use a given `replication_factor` for each DC.

As Avi noted elsewhere, the tablet count should be actually specified per table. It's only per keyspace now because we need whole keyspace to be marked as tablet-based somehow. So want we really want to set is the "tablets: true" option, not "initial_tablets". 

The optimal value of initial tablets chosen by the system depends on current topology when tablets are allocated, and tablets are allocated not when the keyspace is created, but when a table is created. So we should choose it when creating a table.

Reply:
The above is in e.g. scylla-driver-3.26.0.zip/cassandra/metadata.py

```python
class ReplicationFactor(object):
    ...
    @staticmethod
    def create(rf):
        """
        Given the inputted replication factor string, parse and return the ReplicationFactor instance.
        """
        transient_replicas = None
        try:
            all_replicas = int(rf)
        except ValueError:
            try:
                rf = rf.split('/')
                all_replicas, transient_replicas = int(rf[0]), int(rf[1])
            except Exception:
                raise ValueError("Unable to determine replication factor from: {}".format(rf))

        return ReplicationFactor(all_replicas, transient_replicas)

```

Reply:
Yes, both "tablets" and the previous "initial-tablets" will be handled as if they are data center names if you stick them inside the topology strategy. Maybe it can be outside?
Alternatively, we can use "tablets: 0" and "tablets: 1". Ugly, but will work :-( Just not if you have a data center called "tablets"...

Reply:
Maybe it's better to have explicit dedicated replication strategy name for tablets?
And an optional quirk in prepare_options() that would overwrite NetworkTopologyReplicationStrategy with TabletsReplicationStrategy

Reply:
> Just not if you have a data center called "tablets"...

... or "initial_tablets" :)

Reply:
> Maybe it's better to have explicit dedicated replication strategy name for tablets? And an optional quirk in prepare_options() that would overwrite NetworkTopologyReplicationStrategy with TabletsReplicationStrategy

That doesn't seem to save us work, since then you need to teach drivers about this new replication strategy. You can as well teach the drivers about tablet options.

Another reasons not to introduce separate RS for tablets is that tablets is orthogonal to RS spirit, you could make other strategies tablet-based too.

Reply:
Alternatives:

 - a new WITH tablets = boolean_constant clause
 - use `$` - `tablets$enabled`, `tablets$initial_count` (the latter will go away, it belongs on the table level).

Reply:
> Alternatives:
> 
> * a new WITH tablets = boolean_constant clause

That's cleaner, but it means we need to extend schema tables. This means we need to update tools which work with schema sstables. Did we phase-out java tools already?

> * use `$` - `tablets$enabled`, `tablets$initial_count` (the latter will go away, it belongs on the table level).

Reply:
Can it _not_ be the part of DDL? In that case we can put the bit into `system_schema.scylla_keyspaces`, this would solve the [boot-time schema merging issue](https://github.com/scylladb/scylladb/pull/16364#discussion_r1425388543)